File: sdk/typescript/src/telemetry/verifier.ts
import { Detector, defaultDetector } from './detectors';
import { ProcessedTelemetryEvent } from './client-types';

export interface VerificationViolation {
  event: string;
  path: string;
  message: string;
}

export interface VerificationResult {
  valid: boolean;
  violations: VerificationViolation[];
}

const REDACTED_VALUE = '[REDACTED]';

const getValueAtPath = (attributes: Record<string, unknown>, path: string): unknown => {
  const segments = path.replace(/\[(\d+)\]/g, '.$1').split('.').filter(Boolean);
  let current: unknown = attributes;
  for (const segment of segments) {
    if (current === null || current === undefined) {
      return undefined;
    }
    if (Array.isArray(current)) {
      const index = Number(segment);
      current = Number.isNaN(index) ? undefined : current[index];
    } else if (typeof current === 'object') {
      current = (current as Record<string, unknown>)[segment];
    } else {
      return undefined;
    }
  }
  return current;
};

const traverse = (
  value: unknown,
  path: string,
  visit: (path: string, leaf: string) => void,
): void => {
  if (value === null || value === undefined) {
    return;
  }
  if (Array.isArray(value)) {
    value.forEach((item, index) => traverse(item, `${path}[${index}]`, visit));
    return;
  }
  if (typeof value === 'object') {
    Object.entries(value as Record<string, unknown>).forEach(([key, child]) =>
      traverse(child, path ? `${path}.${key}` : key, visit),
    );
    return;
  }
  if (typeof value === 'string') {
    visit(path, value);
  }
};

export class TelemetryVerifier {
  constructor(private readonly detector: Detector = defaultDetector) {}

  verify(events: ProcessedTelemetryEvent[]): VerificationResult {
    const violations: VerificationViolation[] = [];

    for (const event of events) {
      const redactionEntries = event.metadata.redactionMap;

      for (const dropped of event.metadata.droppedFields) {
        const value = getValueAtPath(event.attributes, dropped);
        if (value !== undefined) {
          violations.push({
            event: event.name,
            path: dropped,
            message: 'denied field still present',
          });
        }
      }

      Object.entries(redactionEntries).forEach(([path, entry]) => {
        const value = getValueAtPath(event.attributes, path as string);
        switch (entry.action) {
          case 'redact':
            if (value !== REDACTED_VALUE) {
              violations.push({ event: event.name, path, message: 'redacted field not sanitized' });
            }
            break;
          case 'hash':
            if (typeof value !== 'string' || !value.startsWith('hash:')) {
              violations.push({ event: event.name, path, message: 'hashed field missing hash prefix' });
            }
            break;
          case 'deny':
            if (value !== undefined) {
              violations.push({ event: event.name, path, message: 'denied field serialized' });
            }
            break;
          default:
            break;
        }
      });

      traverse(event.attributes, '', (path, value) => {
        const findings = this.detector.detect(value);
        if (findings.length > 0) {
          const entry = redactionEntries[path];
          if (!entry || (entry.action !== 'redact' && entry.action !== 'hash' && entry.action !== 'deny')) {
            violations.push({
              event: event.name,
              path,
              message: 'pii detected without protective action',
            });
          }
        }
      });
    }

    return { valid: violations.length === 0, violations };
  }
}

File: sdk/python/maestro_sdk/telemetry/verifier.py
from __future__ import annotations

from dataclasses import dataclass
from typing import List, Sequence

from .detectors import Detector, default_detector
from .types import ProcessedTelemetryEvent, TelemetryMetadata

_REDACTED_VALUE = "[REDACTED]"


def _value_at_path(attributes: dict[str, object], path: str) -> object | None:
    if not path:
        return attributes
    segments = [segment for segment in path.replace("[", ".").replace("]", "").split(".") if segment]
    current: object = attributes
    for segment in segments:
        if isinstance(current, list):
            index = int(segment)
            if index >= len(current):
                return None
            current = current[index]
        elif isinstance(current, dict):
            if segment not in current:
                return None
            current = current[segment]
        else:
            return None
    return current


def _traverse(value: object, path: str, collector: List[tuple[str, str]]) -> None:
    if value is None:
        return
    if isinstance(value, list):
        for index, item in enumerate(value):
            _traverse(item, f"{path}[{index}]", collector)
        return
    if isinstance(value, dict):
        for key, item in value.items():
            child = f"{path}.{key}" if path else key
            _traverse(item, child, collector)
        return
    if isinstance(value, str):
        collector.append((path, value))


@dataclass
class VerificationViolation:
    event: str
    path: str
    message: str


@dataclass
class VerificationResult:
    valid: bool
    violations: List[VerificationViolation]


class TelemetryVerifier:
    def __init__(self, detector: Detector | None = None) -> None:
        self._detector = detector or default_detector

    def verify(self, events: Sequence[ProcessedTelemetryEvent]) -> VerificationResult:
        violations: List[VerificationViolation] = []

        for event in events:
            metadata: TelemetryMetadata = event.metadata
            for dropped in metadata.dropped_fields:
                if _value_at_path(event.attributes, dropped) is not None:
                    violations.append(
                        VerificationViolation(event.name, dropped, "denied field still present"),
                    )

            for path, entry in metadata.redaction_map.items():
                value = _value_at_path(event.attributes, path)
                if entry.action == 'redact':
                    if value != _REDACTED_VALUE:
                        violations.append(
                            VerificationViolation(event.name, path, "redacted field not sanitized"),
                        )
                elif entry.action == 'hash':
                    if not isinstance(value, str) or not value.startswith('hash:'):
                        violations.append(
                            VerificationViolation(event.name, path, "hashed field missing hash prefix"),
                        )
                elif entry.action == 'deny':
                    if value is not None:
                        violations.append(
                            VerificationViolation(event.name, path, "denied field serialized"),
                        )

            collected: List[tuple[str, str]] = []
            _traverse(event.attributes, "", collected)
            for path, raw_value in collected:
                findings = self._detector.detect(raw_value)
                if findings:
                    entry = metadata.redaction_map.get(path)
                    if entry is None or entry.action not in {'redact', 'hash', 'deny'}:
                        violations.append(
                            VerificationViolation(event.name, path, "pii detected without protective action"),
                        )

        return VerificationResult(valid=not violations, violations=violations)
